### ISSUE TITLE
training_epoch_end() only gives outputs from one optimizer when multiple optimizers are being used

### ISSUE THREAD

**Issue by import-antigravity** [PyTorchLightning/pytorch-lightning#2687]
I'm making an abstract GAN class, where I have the following code:

```python
    def configure_optimizers(self):
        return self.g_optimizer, self.d_optimizer

    def training_step(self, batch: Tuple[Tensor, Tensor], batch_idx, optimizer_idx) -> Dict:
        X, _ = batch

        batch_size = X.shape[0]

        z = self.Z.sample((batch_size,))
        if self.on_gpu:
            z = z.cuda(X.device)

        g_X = self.G(z)

        if optimizer_idx == 0:
            # Train generator
            d_z = self.D(g_X)

            g_loss = self.generator_loss(d_z)
            step_dict = {'loss': g_loss, 'g_loss': g_loss}
            return step_dict
        if optimizer_idx == 1:
            # Train discriminator
            d_x = self.D(X)
            d_z = self.D(g_X)

            d_loss = self.discriminator_loss(d_x, d_z)
            step_dict = {'loss': d_loss, 'd_loss': d_loss}
            return step_dict
```

In `training_epoch_end()`, `outputs` only contains the discriminator losses. Is this a bug, or am I doing something wrong?

**Comment by rohitgr7** [comment]
With `n` optimizers training_step function is called `n` times. Since here n=2 so after training_step the final batch_output are always `{'d_loss': d_loss}`. A hack I would suggest for now is:
```python
if optimizer_idx == 0:
    self.train_out = {'g_loss': g_loss}
    step_dict = {'loss': g_loss}
    return step_dict
elif optimizer_idx == 1:
    self.train_out.update({'d_loss': d_loss})
    step_dict = {'loss': d_loss, 'losses': self.train_out}
    return step_dict
```
or, simply
```python
if optimizer_idx == 0:
    self.g_loss = g_loss
    step_dict = {'loss': g_loss}
    return step_dict
elif optimizer_idx == 1:
    step_dict = {'loss': d_loss, 'g_loss': self.g_loss, 'd_loss': d_loss}
    return step_dict
```

There is something new in pl `TrainResult`, not sure if that will make this easy to handle, but for now, this can be a workaround.

**Comment by import-antigravity** [comment]
There's no `train_out` attribute

**Comment by import-antigravity** [comment]
I can try that, but I think the 'loss' key is getting overwritten instead of added to

**Comment by import-antigravity** [comment]
Okay, that made `g_loss` and `d_loss` save, but `loss` is still the discriminator loss.

**Comment by rohitgr7** [comment]
Yeah, loss will remain as discriminator loss due the the replacement happening here, since the key name is same in both the cases. But there will be a backward call for each loss, the gradient update is not a problem here.

**Comment by import-antigravity** [comment]
Okay. Let me make sure that the generator is actually being updated. Is this a bug, or is this how GANs should be trained at this point?

**Comment by rohitgr7** [comment]
It has nothing to do with how Gans should be trained. It's just how lightning handles training step in case of multiple optimizers. The core backprop and update is happening correctly, no issue with that, but maybe in future this thing with training step will be fixed if required.

**Comment by import-antigravity** [comment]
Okay, I'll keep it open for now because I think there should be a much better way to do this
